### PR TITLE
Improves in unsafe memory

### PR DESCRIPTION
**Pull request by gortiz**
This PR continues what has been done in #12794.

### CODE REVIEW

**Review by jackjlli — APPROVED** [review]
LGTM.

**Comment by siddharthteotia** [comment]
I feel we can remove the method level `synchronize` and declare the field as `volatile` ?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/DirectMemory.java [hunk 1]
```diff
@@ -56,8 +56,12 @@ public void flush() {
   @Override
   public synchronized void close() {
     if (!_closed) {
-      Unsafer.UNSAFE.freeMemory(_address);
-      _closed = true;
+      synchronized (this) {
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/DirectMemory.java:59]: This is not necessary. The whole `close()` is synchronized

Comment (gortiz) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/DirectMemory.java:59]: You are right, but we can remove that first synchronized now, right?

**Comment by gortiz** [comment]
> I feel we can remove the method level `synchronize` and declare the field as `volatile` ?

I think that would be safe as well, although the reasoning may be a bit more complex. Given the catastrophic impact of being wrong, I think it is safer to use synchronize. At least, while we are not using virtual threads, it should be fairly cheap.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/DirectMemory.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/unsafe/DirectMemory.java:59]: This method is almost always invoked only once, so the double-checked locking is not helping much here. Even for singleton class access, I've read that for modern JVM the overhead of locking is negligible, and people are suggesting using global locking for simplicity

**Pull request closed without merging**